### PR TITLE
update writers for consistency

### PR DESCRIPTION
**Pull request by llllllllll**
I am splitting this out of the quandl data loader work.

### CODE REVIEW

File: zipline/algorithm.py [hunk 1]
```diff
@@ -559,6 +569,17 @@ def run(self, data=None, overwrite_sim_params=True):
                 data = data.swapaxes(0, 2)
 
             if isinstance(data, pd.Panel):
+                # For compatibility with existing examples allow start/end
+                # to be inferred.
+                if overwrite_sim_params:
```
Comment (richafrank) [inline zipline/algorithm.py]: Why this move?

File: zipline/algorithm.py [hunk 2]
```diff
@@ -624,10 +636,30 @@ def _write_and_map_id_index_to_sids(self, identifiers, as_of_date):
                 asset = self.asset_finder.retrieve_asset(sid=identifier,
                                                          default_none=True)
             if asset is None:
-                identifiers_to_build.append(identifier)
-
-        self.trading_environment.write_data(
-            equities_identifiers=identifiers_to_build)
+                try:
+                    sid = op.index(identifier)
```
Comment (richafrank) [inline zipline/algorithm.py]: Did we have this logic somewhere else before?

File: zipline/assets/asset_db_migrations.py [hunk 3]
```diff
@@ -68,13 +71,46 @@ def _pragma_foreign_keys(connection, on):
...
 
-def _downgrade_v1_to_v0(op, version_info_table):
+# This dict contains references to downgrade methods that can be applied to an
+# assets db. The resulting db's version is the key.
+# e.g. The method at key '0' is the downgrade method from v1 to v0
+_downgrade_methods = {}
+
+
+def downgrades(src):
+    """Decorator for marking that a method is a downgrade to a version to the
```
Comment (richafrank) [inline zipline/assets/asset_db_migrations.py]: typo? "a downgrade to a version to the previous version"

File: zipline/assets/asset_db_migrations.py [hunk 4]
```diff
@@ -68,13 +71,46 @@ def _pragma_foreign_keys(connection, on):
...
+    decorator : callable[(callable) -> callable]
+        The decorator to apply.
+    """
+    def _(f):
+        destination = src - 1
+
+        @do(op.setitem(_downgrade_methods, destination))
+        @wraps(f)
+        def wrapper(op, version_info_table):
+            version_info_table.delete().execute()  # clear the version
```
Comment (richafrank) [inline zipline/assets/asset_db_migrations.py]: Does alembic not manage this for us?

File: zipline/assets/asset_db_migrations.py [hunk 5]
```diff
@@ -126,12 +159,50 @@ def _downgrade_v2_to_v1(op, version_info_table):
...
-    0: _downgrade_v1_to_v0,
-    1: _downgrade_v2_to_v1,
-}
+@downgrades(3)
+def _downgrade_v3(op):
+    """
+    Downgrade assets db by adding a not null constraint on
+    ``equities.first_traded``
+    """
+    op.create_table(
```
Comment (richafrank) [inline zipline/assets/asset_db_migrations.py]: What about futures?

File: zipline/assets/asset_writer.py [hunk 6]
```diff
@@ -171,6 +166,27 @@ def _generate_output_dataframe(data_subset, defaults):
...
+    dt_series : pd.Series
+        The timeseries to convert.
+
+    Returns
+    -------
+    idx : pd.Int64Index
+        The index converted to nanoseconds since the epoch.
+    """
+    index = pd.to_datetime(dt_series.values)
+    if index.tzinfo is None:
```
Comment (richafrank) [inline zipline/assets/asset_writer.py]: How does this perform compared to the try-except?

File: zipline/assets/asset_writer.py [hunk 7]
```diff
@@ -216,171 +232,174 @@ def write_version_info(version_table, version_value):
...
-    @abstractmethod
-    def defaultval(self):
-        raise NotImplementedError
-
-    def write_all(self,
-                  engine,
-                  allow_sid_assignment=True):
-        """ Write pre-supplied data to SQLite.
+    def __init__(self, engine, allow_sid_assignment=True):
+        if isinstance(engine, str):
```
Comment (richafrank) [inline zipline/assets/asset_writer.py]: string_types?

File: zipline/assets/assets.py [hunk 8]
```diff
@@ -101,6 +101,8 @@ class AssetFinder(object):
     PERSISTENT_TOKEN = "<AssetFinder>"
 
     def __init__(self, engine):
+        if isinstance(engine, str):
```
Comment (richafrank) [inline zipline/assets/assets.py]: string_types

File: zipline/assets/assets.py [hunk 9]
```diff
@@ -288,6 +290,30 @@ def retrieve_equities(self, sids):
         """
         return self._retrieve_assets(sids, self.equities, Equity)
 
+    def _make_info(attr):
+        def _(self):
+            return pd.DataFrame.from_records(
+                map(
+                    methodcaller('to_dict'),
```
Comment (richafrank) [inline zipline/assets/assets.py]: A list comp might be clearer here, but just a question of style.

File: zipline/assets/assets.py [hunk 10]
```diff
@@ -288,6 +290,30 @@ def retrieve_equities(self, sids):
...
+                    self.retrieve_all(getattr(self, attr)),
+                ),
+                index='sid',
+            )
+
+        return _
+
+    equities_info = property(
+        _make_info('equities_sids'),
+        doc='A dataframe that could be fed to an AssertDBWriter for equities'
```
Comment (richafrank) [inline zipline/assets/assets.py]: typo here and below "AssertDBWriter" -> "AssetDBWriter"

File: zipline/assets/assets.py [hunk 11]
```diff
@@ -288,6 +290,30 @@ def retrieve_equities(self, sids):
         """
         return self._retrieve_assets(sids, self.equities, Equity)
 
+    def _make_info(attr):
```
Comment (richafrank) [inline zipline/assets/assets.py]: Is this in the class namespace to put it near the call sites?

File: zipline/assets/assets.py [hunk 12]
```diff
@@ -684,12 +710,30 @@ def lookup_expired_futures(self, start, end):
 
         return sids
 
-    @property
-    def sids(self):
-        return tuple(map(
-            itemgetter('sid'),
-            sa.select((self.asset_router.c.sid,)).execute().fetchall(),
-        ))
+    def _make_sids(tblattr):
```
Comment (richafrank) [inline zipline/assets/assets.py]: Can we find a better name for "tblattr"?

File: zipline/assets/assets.py [hunk 13]
```diff
@@ -684,12 +710,30 @@ def lookup_expired_futures(self, start, end):
...
+        def _(self):
+            return tuple(map(
+                itemgetter('sid'),
+                sa.select((
+                    getattr(self, tblattr).c.sid,
```
Comment (richafrank) [inline zipline/assets/assets.py]: Where do we add these table attributes to the instance now?

File: zipline/assets/assets.py (hunk 11 above)
Comment (llllllllll) [inline zipline/assets/assets.py]: yeah, when I make a function for generating methods I like to keep it next to the use and del it later. I use this pattern in the fixtures for making a lot of "return None" functions. This is harder to understand when it is not in the class imho.

File: zipline/data/minute_bars.py [hunk 14]
```diff
@@ -528,11 +528,16 @@ def write_cols(self, sid, dts, cols):
...
-        open_col[dt_ixs] = (cols['open'] * ohlc_ratio).astype(np.uint32)
-        high_col[dt_ixs] = (cols['high'] * ohlc_ratio).astype(np.uint32)
-        low_col[dt_ixs] = (cols['low'] * ohlc_ratio).astype(np.uint32)
-        close_col[dt_ixs] = (cols['close'] * ohlc_ratio).astype(
-            np.uint32)
+
+        def convert_col(col):
+            """Adapt float column into a uint32 column.
+            """
+            return (np.nan_to_num(col) * ohlc_ratio).astype(np.uint32)
```
Comment (richafrank) [inline zipline/data/minute_bars.py]: Why did we add this?

File: zipline/data/us_equity_pricing.py [hunk 15]
```diff
@@ -48,40 +46,36 @@
...
-US_EQUITY_PRICING_BCOLZ_COLUMNS = [
+US_EQUITY_PRICING_BCOLZ_COLUMNS = (
     'open', 'high', 'low', 'close', 'volume', 'day', 'id'
-]
-SQLITE_ADJUSTMENT_COLUMNS = frozenset(['effective_date', 'ratio', 'sid'])
+)
 SQLITE_ADJUSTMENT_COLUMN_DTYPES = {
     'effective_date': integer,
-    'ratio': floating,
+    'ratio': float,
```
Comment (richafrank) [inline zipline/data/us_equity_pricing.py]: What's the difference here?

File: zipline/data/us_equity_pricing.py [hunk 16]
```diff
@@ -175,19 +218,40 @@ def write(self, filename, calendar, assets, show_progress=False):
...
-                return self._write_internal(filename, calendar, pbar_iterator)
-        return self._write_internal(filename, calendar, _iterator)
+        ctx = maybe_show_progress(
+            map(_snd_to_ctable, data),
+            show_progress=show_progress,
+            item_show_func=self.progress_bar_item_show_func,
+            label=self.progress_bar_message,
+            length=len(assets) if assets is not None else None,
+        )
+        with ctx as it:
```
Comment (richafrank) [inline zipline/data/us_equity_pricing.py]: Does the name `it` have a meaning here?

Comment (richafrank) [inline zipline/data/us_equity_pricing.py]: Ah, found it (iterator).

File: zipline/data/us_equity_pricing.py [hunk 17]
```diff
@@ -205,6 +269,15 @@ def _write_internal(self, filename, calendar, iterator):
         }
 
         earliest_date = None
+        calendar = self._calendar
+
+        if assets is not None:
+            @apply
```
Comment (richafrank) [inline zipline/data/us_equity_pricing.py]: Might be clearer to people to call the function at the call site instead?

Comment (llllllllll) [inline zipline/data/us_equity_pricing.py]: in the other branch iterator is already the iterator, here I just wanted to make another wrapper around that if we are passed assets.

File: zipline/data/us_equity_pricing.py [hunk 18]
```diff
@@ -401,36 +412,58 @@ class BcolzDailyBarReader(DailyBarReader):
...
+        return {
             int(asset_id): start_index
-            for asset_id, start_index in iteritems(table.attrs['first_row'])
+            for asset_id, start_index in iteritems(
+                self._table.attrs['first_row'],
+            )
         }
-        self._last_rows = {
+
+    @lazyval
```
Comment (richafrank) [inline zipline/data/us_equity_pricing.py]: What's the advantage of lazyval versus the instance attribute in these cases?

File: zipline/data/us_equity_pricing.py [hunk 19]
```diff
@@ -841,14 +862,12 @@ def calc_dividend_ratios(self, dividends):
...
-        calendar = self._calendar
-
         effective_dates = full(len(amounts), -1, dtype=int64)
-
+        calendar = self._calendar
         for i, amount in enumerate(amounts):
             sid = sids[i]
             ex_date = ex_dates[i]
-            day_loc = calendar.get_loc(ex_date)
+            day_loc = calendar.get_loc(ex_date, method='bfill')
```
Comment (richafrank) [inline zipline/data/us_equity_pricing.py]: What was this for?

File: zipline/finance/trading.py [hunk 20]
```diff
@@ -114,11 +124,11 @@ def __init__(
         if isinstance(asset_db_path, string_types):
             asset_db_path = 'sqlite:///%s' % asset_db_path
             self.engine = engine = create_engine(asset_db_path)
-            AssetDBWriterFromDictionary().init_db(engine)
         else:
             self.engine = engine = asset_db_path
 
         if engine is not None:
+            AssetDBWriter(engine).init_db()
```
Comment (richafrank) [inline zipline/finance/trading.py]: Does the difference in location mean that it's more robust to a db already being initialized?

File: zipline/pipeline/loaders/synthetic.py [hunk 21]
```diff
@@ -195,9 +192,29 @@ def _bool_values(self, shape):
         return self.state.randn(*shape) < 0
 
 
-class SyntheticDailyBarWriter(BcolzDailyBarWriter):
+OHLCV = ('open', 'high', 'low', 'close', 'volume')
```
Comment (richafrank) [inline zipline/pipeline/loaders/synthetic.py]: Worth making this `OHLC + ('volume',)`?

File: zipline/pipeline/loaders/synthetic.py [hunk 22]
```diff
@@ -221,50 +238,50 @@ class SyntheticDailyBarWriter(BcolzDailyBarWriter):
...
-        ]
+        dates = calendar[calendar.slice_indexer(
+            asset_start(asset_info, asset_id),
+            asset_end(asset_info, asset_id),
+        )]
 
         data = full(
             (len(dates), len(US_EQUITY_PRICING_BCOLZ_COLUMNS)),
-            asset_id * (100 * 1000),
+            asset_id * 10000,
```
Comment (richafrank) [inline zipline/pipeline/loaders/synthetic.py]: Any reason for the 10x decrease here and below?

File: zipline/assets/assets.py (hunk 12 above)
Comment (llllllllll) [inline zipline/assets/assets.py]: how is `table_attr_name`?

File: zipline/data/minute_bars.py (hunk 14 above)
Comment (llllllllll) [inline zipline/data/minute_bars.py]: I had run into the same bug as Eddie and we both had a fix. This just pulls the logic into a function and shares it for all the columns.

File: zipline/data/us_equity_pricing.py (hunk 18 above)
Comment (llllllllll) [inline zipline/data/us_equity_pricing.py]: The calendar is not always populated at construction time.

File: zipline/finance/trading.py (hunk 20 above)
Comment (llllllllll) [inline zipline/finance/trading.py]: This was to make sure that we are always working with an initialized db. Now that I think about this, this should probably just be part of the constructor. How do you feel about that?

File: zipline/data/us_equity_pricing.py (hunk 19 above)
Comment (llllllllll) [inline zipline/data/us_equity_pricing.py]: if you have a dividend that is payed out of a non-trading day it should use the last trading day's price

File: zipline/pipeline/loaders/synthetic.py (hunk 22 above)
Comment (llllllllll) [inline zipline/pipeline/loaders/synthetic.py]: this is a bug.

**Comment by llllllllll** [comment]
closing in favor of #1108

**Pull request closed without merging**